bigrig/config.py

```diff
@@ -0,0 +1,202 @@
+from packaging.requirements import Requirement
+
+import yaml
+import jsonschema
+
+CREDENTIALS_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "username": {"type": "string", "minLength": 1},
+        "password": {"type": "string"},
+    },
+    "required": ["username", "password"],
+    "additionalProperties": False,
+}
+
+CONFIG_SCHEMA = {
+    "type": "object",
+    "definitions": {
+        "origin": {
+            "type": "object",
+            "description": "this is basically big public Pypi, where we get initial sources from. It may be external",
+            "properties": {
+                "location": {
+                    "type": "string",
+                    "description": "Local FS folder or url to pypi simple compatible repo",
+                },
+                "credentialsPath": {
+                    "type": "string",
+                    "description": "Path to YAML file containing CREDENTIALS_SCHEMA",
+                },
+            },
+            "required": ["location"],
+            "additionalProperties": False,
+        },
+        "source": {
+            "type": "object",
+            "description": "where we put sources fetched from origin for internal keeping. Need only one of those",
+            "properties": {
+                "location": {
+                    "type": "string",
+                    "description": "url to pypi simple compatible repo",
+                }
+            },
+            "required": ["location"],
+            "additionalProperties": False,
+        },
+        "target": {
+            "type": "object",
+            "properties": {
+                "variables": {
+                    "type": "object",
+                    "description": "Variables to be forwarded to the dockerfile template and used inside",
+                },
+                "location": {
+                    "type": "string",
+                    "description": "url to pypi simple compatible repo",
+                },
+            },
+            "required": ["variables", "location"],
+            "additionalProperties": False,
+        },
+    },
+    "properties": {
+        "origin": {"$ref": "#/definitions/origin"},
+        "source": {"$ref": "#/definitions/source"},
+        "targets": {
+            "type": "object",
+            "description": "where we put the built things. As many as we have arches + pythonversions ( or maybe just arches? You can have several pythonversions coexist in one repo)",
+            "patternProperties": {"^[a-z0-9_]+$": {"$ref": "#/definitions/target"}},
+        },
+    },
+}
+
+
+@dataclass
+class Origin:
+    location: str
+    credentials: t.Optional["OriginCredentials"]
+
+    @classmethod
+    def from_dict(cls, blob: t.Dict) -> "Origin":
+        return cls(
+            location=blob["location"],
+            credentials=OriginCredentials.from_path(path=blob.get("credentialsPath")),
+        )
+
+
+@dataclass
+class OriginCredentials:
```

Comment:
I see only origin seems to have credentials, and naming reflects that.
All the repos should have it actually. In fact, the way it works right now is  you need password only to upload, so origin is the only one that does not require credentials :)

Reply:
This is what we discussed initially, but I think it just makes sense to allow having credentials in every origin,source,target entry, I'll review and update the design doc accordingly :)